x-pack/plugins/enterprise_search/server/lib/enterprise_search_request_handler.ts

```diff
@@ -126,6 +130,36 @@ export class EnterpriseSearchRequestHandler {
     };
   }
 
+  /**
+   * This path helper is similar to React Router's generatePath, but much simpler &
+   * does not use regexes. It enables us to pass a static '/foo/:bar/baz' string to
+   * createRequest({ path }) and have :bar be automatically replaced by the value of
+   * request.params.bar.
+   * It also (very importantly) wraps all URL request params with encodeURIComponent(),
+   * which is an extra layer of encoding required by the Enterprise Search server in
+   * order to correctly & safely parse user-generated IDs with special characters in
+   * their names - just encodeURI alone won't work.
+   */
+  encodePathParams(path: string, params: UnknownObject) {
+    const hasParams = path.includes(':');
+    if (!hasParams) {
+      return path;
+    } else {
+      return path
+        .split('/')
+        .map((pathPart) => {
+          const isParam = pathPart.startsWith(':');
+          if (!isParam) {
+            return pathPart;
```

Comment:
FYI: I opted to leave non dynamic parts of the URL unencoded. (i.e., in the URL `/hello/:world`, hello is left as-is and `world` is encoded.

I figured that any non-`:` URL is coming from us, statically, set directly in the server side code, and is therefore going to be safe because it's very unlikely that we write a bad URL. That being said  I'm happy to wrap this in `encodeURIComponent()` as well if people think it's helpful, just let me know.

Reply:
Can you explain this again, I'm kind of confused, I'm missing something about why `encodeURIComponent(pathPart) ` would be optional I think

Reply:
Sorry, probably should have used a real URL instead of a dummy one. Example real Enterprise Search URL from this PR that we're sending:

```js
path: `/as/engines/:engineName/analytics/query/:query`,
```

Right now only the dynamic user-generated `engineName` and `query` vars are getting encoded. I figured I didn't need to bother with the static `/as/engines/.../analytics/query/...` part of the URL because it's static and we're in control of it (not user generated).

The worse case scenario to bother `encodeURI`ing the static portions of the URL for is if in the future some day some AS dev is alseep or accidentally writes some URL like `/as/eng i nes/.../whatev[]er` or simliar, but I thought that a bit improbable, so skipped encoding 🤷 That being said if we want to cover our bases, I'm fine w/ it